### DISCUSSION TITLE
Server-side only data fetching (like Next.js' `getServerSideProps()`)

### DISCUSSION THREAD

**Discussion by brillout** [nuxt/nuxt#15924] [Questions] [answered]
Next.js has `getServerSideProps()` which is not isomorphic but server-side only: it's always called on the server-side. (Even when the user client-side navigates to that page.)

This allows users to directly use SQL/ORM queries in order to fetch the page's initial data.

```js
// Next.js
// pages/*.js

export async function getServerSideProps() {
  // `getServerSideProps()` is always called on the server-side.
  // We can use any SQL/ORM query here.
}
```

For more context about what I mean, see [Telefunc > Server-Side Rendering > SSR frameworks](https://telefunc.com/ssr#ssr-frameworks).

Am I right to assume that Nuxt v3 doesn't have any data fetching hook that is not isomorphic but server-side only?

The [Vue 3 - Data Fetching](https://v3.nuxtjs.org/docs/usage/data-fetching) docs doesn't mention that.

(I'm writing Nuxt docs for [Telefunc](https://telefunc.com/) and I'd like to write proper Nuxt + Telefunc usage recommendation.)

**Comment by mikield** [comment]
You can check if the composable is called on the server or not.
`process.server` Is holding boolean for that.

For example, I have tested an SQLite integration using composables.
![image](https://user-images.githubusercontent.com/4073672/153544545-d73b4763-9968-45f9-b89f-fed2d42c975e.png)

And got the results:
```bash
1: Ipsum 0
2: Ipsum 1
3: Ipsum 2
4: Ipsum 3
5: Ipsum 4
6: Ipsum 5
7: Ipsum 6
8: Ipsum 7
9: Ipsum 8
10: Ipsum 9
```

**Reply by mikield** [reply]
Using that with `useState` will perform that action that you need.

`useState` as said in documentation is shared between server and client.
So if you put something on the server - your client will get that data 👍🏼

**Reply by brillout** [reply]
Isn't that only half the story though? What happens when the user navigates from another page to that page (i.e. browser-side navigation); the SQL will not be run, right? This means that SQL cannot be used to fetch the initial data of the page.

**Reply by mikield** [reply]
Client side navigation is made on clients side. It's technically is not possible to use socket related things in browser. 

If nuxt will allow severOnly mode, users navigation will be performed with page refresh and SSR calling your database on sever.

**Reply by brillout** [reply]
That's actually my point: seems like Nuxt does not have any data fetching hook that is not isomorphic but server-side only, i.e. always called on the server-side and never in the browser.

> severOnly mode

I searched in the docs of both Nuxt v2 and Nuxt v3 and couldn't find `severOnly`, not sure what you are talking about here.

Downvoting your answer since it doesn't answer my original question. Thanks for the answer though.

I edited the question, it wasn't very clear &mdash; sorry for that.

**Reply by mikield** [reply]
Sorry, I was not very clear, I meant that: If Nuxt will bring some serverOnly mode, like universal and spa in Nuxt2 - that will resolve your issue :)

By the time - your initial request will be rendered by the server, you (theoretically) can perform calls to DB and set state.

**Reply by brillout** [reply]
As explained in my previous answer, I don't believe this would resolve the issue. Any SQL/ORM query will choke upon page navigation but is needed for the navigated page to render.

**Comment by IvanBernatovic** [comment, marked as answer]
> Am I right to assume that Nuxt v3 doesn't have any data fetching hook that is not isomorphic but server-side only?

@brillout You are, for better or worse, there's no such thing as `getServerSideProps` in Nuxt. I think this is the biggest difference between the two. I kind of miss it in Nuxt because with `getServerSideProps` you don't need to implement a separate API code, you can call DB directly and don't need to worry about permissions or security implications.

I still think Nuxt is more elegant with async fetch and the ability to change isomorphic/SPA modes without changing any code (given you follow the guidelines).

**Reply by jeffydc** [reply]
So Nuxt is making one more additional request on server? For example,
- on the initial first request from browser (browser -> server, +1 request)
- Nuxt has to make another request to the api endpoint to get the data (server -> endpoint, +1 request)
- which in turn can make a request to the DB (+1 request).

**Reply by danielroe** [reply]
In fact there isn't an additional request. See https://v3.nuxtjs.org/concepts/server-engine#direct-api-calls for details on what happens.

**Reply by jeffydc** [reply]
Thanks for the clarification.

**Reply by brillout** [reply]
@danielroe While you are here, so you can confirm that there is no Nuxt hook that is guaranteed to always run in Node.js? I'll write the Telefunc's Nuxt docs accordingly. (In case you are curious: https://telefunc.com/ which actually plays really nicely with Nuxt, espcially given that Nuxt doesn't have such server-side only hook. There is already a Nuxt v2 example and I'll implement a Nuxt v3 example soon :-).)

**Reply by danielroe** [reply]
what do you mean by a hook 'guaranteed to always run in node.js'?

* anything you wrap in `if (process.server)` will run only on server-side, for example in the `created()` or `setup()` lifecycle hooks of a component, or even in ambient javascript
* server-only plugins (e.g. `~/plugins/test.server.ts`) will only run on server-side

**Reply by IvanBernatovic** [reply]
@danielroe He (@brillout ) is asking does Nuxt 3 have a strict equivalent to Next's `getServerSideProps` function. I feel that many Nuxt devs don't understand what is `getServerSideProps` because it doesn't exist in Nuxt. I'll try to describe it quickly here so you can provide a definitive yes/no answer to @brillout.

You define and export `getServerSideProps` in the same file/component as your page component. The return value of `getServerSideProps` should be an object and all of the properties are passed to the page as props. If the page component file has this function defined Next will always run this function on the server and it will never end up in a client bundle. This means you don't necessarily have to use API routes and you can directly connect to your database, run some queries and return results to the page component as a prop. There are 2 cases (c/p from [Next docs on getServserSideProps](https://nextjs.org/docs/basic-features/data-fetching/get-server-side-props)):

> 1. When you request this page directly, getServerSideProps runs at request time, and this page will be pre-rendered with the returned props
> 2. When you request this page on client-side page transitions through [next/link](https://nextjs.org/docs/api-reference/next/link) or [next/router](https://nextjs.org/docs/api-reference/next/router), Next.js sends an API request to the server, which runs getServerSideProps

So @brillout your initial observation was correct, there's no `getServerSideProps` equivalent in Nuxt.

However, Nuxt 3 introduced [Direct API calls](https://v3.nuxtjs.org/concepts/server-engine#direct-api-calls) which is somewhat similar to getServerSideProps but a bit less elegant IMO. Basically, you can do this:
- use Nuxt's API routes which handle HTTP requests
- you consume those routes in your Nuxt 3 app with the new $fetch function
- but now this $fetch works differently in server-side rendered mode and client-side mode (similar to how next handles `getServerSideProps` differently in a different context):
  1. When you request this page directly, Nuxt will not call your Nuxt API route as HTTP request, it will run the function directly instead ( I guess it's smart enough to realise that this is "internal" route so no need for HTTP requests in the server context)
  2. When you request this page on the client-side page transitions it will make a regular HTTP request

**Reply by brillout** [reply]
Yes that's exactly what I meant.

I think the docs should be updated to clarify that. And maybe mention [Telefunc](https://telefunc.com/) 😉 which is basically a neat way to achieve a `getServerSideProps()`-like DX with Nuxt.

**Reply by blocka** [reply]
however, with telefunc, the code is still in a separate file, right? You still can't drop it into the same file as the component, so not quite the same level of DX.

**Reply by brillout** [reply]
Yes. Although, in principle, it's possible to extend `.vue` files to be able to define telefunctions inside them. PR welcome.

**Reply by brillout** [reply]
E.g. using @galvez's [`acorn-strip-function`](https://github.com/galvez/acorn-strip-function).

**Reply by Youhan** [reply]
I sometimes find Next's `getServerSideProps` that are in the same file, confusing. Not a good DX for me! :)

**Reply by atinux** [reply]
What about https://github.com/antfu/nuxt-server-fn ?

**Comment by galvez** [comment]
FWIW I built **[`acorn-strip-function`](https://github.com/galvez/acorn-strip-function)** with that in mind — that is, being able to export a function from a Vue component (under a vanilla `<script>` section), load it on the server side, use it for data prefetching or registering payload routes, and cleanly remove that function from the client bundle via a Vite plugin. Hope it helps.

**Reply by blocka** [reply]
do you have an example of this?

**Comment by mcrapts** [comment]
Have there been any ideas regarding this subject? I think a compiler macro could easily achieve this, as all the building blocks are already in place. I imagine something like this:

```
const serversideProps = defineServersideProps<any>(() => /* do server-side stuff */)
```
On the server, this function would execute directly. On the client, this function will be served using a server route and consumed using useFetch. This would be very elegant, because you then don't need to explicitly create a server route. It also makes it very easy to reason about your code. I have no idea how easy it is to implement, but from an outsider perspective I feel it would only require a tiny layer of syntactic sugar.

Curious to hear your thoughts.
